src/files/pyfingerprint/pyfingerprint.py

```diff
@@ -673,8 +677,86 @@ def readImage(self):
         else:
             raise Exception('Unknown error '+ hex(receivedPacketPayload[0]))
 
-    ## TODO:
-    ## Implementation of uploadImage()
+    def uploadImage(self, imageLocation):
+    """
+    Upload the image of a finger from host computer.
+
+    @param string imageLocation
+    @return void
+    """
+    locationDirectory = os.path.dirname(imageLocation)
+    if not os.access(locationDirectory, os.R_OK):
+        raise ValueError('The given location directory "' + locationDirectory + '" is not readable!')
+
+    if not os.path.exists(imageLocation):
+        raise ValueError('File "' + imageLocation + '" does not exist!')
+
+    maxPacketSize = self.getMaxPacketSize()
+
+    print("Sending first packet to upload image...") # debug
+    packetPayload = (
+        FINGERPRINT_UPLOADIMAGE,
+    )
+
+    self.__writePacket(FINGERPRINT_COMMANDPACKET, packetPayload)
+
+    ## Get first reply packet
+    receivedPacket = self.__readPacket()
+
+    receivedPacketType = receivedPacket[0]
+    receivedPacketPayload = receivedPacket[1]
+
+    if receivedPacketType != FINGERPRINT_ACKPACKET:
+       raise Exception('The received packet is no ack packet!')
+    # DEBUG: The sensor will send follow-up packets
+    if receivedPacketPayload[0] == FINGERPRINT_OK:
+        pass
+    elif receivedPacketPayload[0] == FINGERPRINT_ERROR_COMMUNICATION:
+        raise Exception('Communication error')
+    elif receivedPacketPayload[0] == FINGERPRINT_ERROR_UPLOADIMAGE:
+        raise Exception('Could not upload image')
+    else:
+        raise Exception('Unknown error ' + hex(receivedPacketPayload[0]))
+
+    # Initialize image library
+    imageToUpload = Image.open(imageLocation)
+    imageArray = numpy.array(imageToUpload)
+
+    maxRowLength = max(map(lambda x: len(x), imageArray))
+
+    print("will start sending data packets")
+
+    if maxRowLength <= maxPacketSize:
+        print("sending 1 row of pixels at a time") # debug
+        i = 1
+        for i in range(0, len(imageArray)-1):
+            self.__writePacket(FINGERPRINT_DATAPACKET, imageArray[i])
+        self.__writePacket(FINGERPRINT_ENDDATAPACKET, imageArray[len(imageArray)-1])
+
+    else:
+        # we need to flatten the list of lists into a single list and divide
+        print("flattening pixels' arrays into a list and sending packets with length: " + str(maxPacketSize)) # debug
+        flattenedList = list(imageArray.flatten())
+        packetNbr = len(flattenedList) / maxPacketSize
```

Comment:
Please avoid abbreviations in code because it may not be clear for others, what you mean. Better `packetCount` or something

Reply:
I used the same nomenclature used in uploadCharacteristics, but sure, I'll change it

Reply:
Yeah the `uploadCharacteristics` method is from a contribution from pre-Github times, we should fix the naming there in the future.